Fix for Android aarch64 socket type

Txuritan opened the pull request:
I separated both functions and cast the u32 to i32 in the bind call.

This is similar to how mio handled the same problem.

Txuritan committed:
Attempted to fix Android socket type

At the moment this is just a duplication of the functions in question.

Not sure how safe this is, in truth the problem could also be fixed in the libc crate.

Fixes alexcrichton/mio-uds#9

For now until a better solution is found.

Txuritan committed:
Forgot to specify the target OS

Txuritan commented:
I could use a if statement using a cfg macro, I was just trying to follow how mio solved it.

Give me a few minutes and I should have a better solution.

Txuritan commented:
So attempting to use if statements with the cfg macro caused type errors, trying to create a separate variable also lead to type errors, the final thing I tried was having a inline cfg if statement where the len variable was passed, this also lead to type errors.

alexcrichton commented:
Oh sorry what I mean is something like:

```rust
#[cfg(foo)]
type dest = i32;
#[cfg(not(foo))]
type dest = u32;

// ...
bind(..., foo as dest);
```

(or something like that)

Txuritan committed:
Changed the function duplication to typedefs

Replaced both instance of 'len as i32' to 'len as Len', the Len typedef is located in the master lib file

Txuritan commented:
I have changed the duplicated functions to a typedef located in lib.rs.

The name can be changed, I just thought it was fitting to name it after what its being applied to.

alexcrichton commented:
Thanks!

NickAtAccuPS commented:
Heads up this PR broke Android Arm builds where socklen_t is [defined as i32](https://github.com/rust-lang/libc/blob/3054d3e0eb272276300e6825ea31eaa1ed8a6f84/src/unix/notbsd/android/b32/mod.rs#L9) and bind et al. have [correct function signatures](https://github.com/rust-lang/libc/blob/3054d3e0eb272276300e6825ea31eaa1ed8a6f84/src/unix/notbsd/android/b32/mod.rs#L201-L217). Since the special case check just applies to aarch64 due to an incorrect function definition I think the correction should be as simple as defining Len as socklen_t in the non-aarch cases.

NickAtAccuPS commented:
No worries, putting together a PR arm.
